torch/export/exported_program.py

```diff
@@ -338,295 +269,139 @@ def _decompose_and_get_gm_with_new_signature_constants(
-            elif spec.kind == InputKind.USER_INPUT:
-                kind = OutputKind.GRADIENT_TO_USER_INPUT
-                target = source
-            else:
-                raise AssertionError(f"Unknown input kind: {spec.kind}")
-            output_specs.append(
-                OutputSpec(
-                    kind,
-                    TensorArgument(name=node.name),
-                    target,
-                )
-            )
+    for node in gm.graph.nodes:
+        # nn_module_stack
+        if node.op not in ["placeholder", "output"]:
+            for key, (fqn, mod_cls) in node.meta["nn_module_stack"].items():
+                if isinstance(mod_cls, type):
+                    node.meta["nn_module_stack"][key] = (
+                        fqn,
+                        mod_cls.__module__ + "." + mod_cls.__qualname__,
+                    )
 
-    assert len(new_placeholders) == len(old_placeholders)
+    # Don't copy over nn_module_stack, stack_trace metadata for params/buffers nodes
+    for metadata in params_buffers_to_node_meta.values():
+        metadata.pop("nn_module_stack", None)
+        metadata.pop("stack_trace", None)
 
-    new_graph_signature = ExportGraphSignature(
-        input_specs=input_specs, output_specs=output_specs
-    )
-    # NOTE: aot_export adds symint metadata for placeholders with int
-    # values; since these become specialized, we replace such metadata with
-    # the original values.
-    # Also, set the param/buffer metadata back to the placeholders.
-    for old_node, new_node in zip(old_placeholders, new_placeholders):
-        if not isinstance(old_node.meta["val"], torch.Tensor):
-            new_node.meta["val"] = old_node.meta["val"]
-
-        if (
-            new_node.target in new_graph_signature.inputs_to_parameters
-            or new_node.target in new_graph_signature.inputs_to_buffers
-        ):
-            for k, v in old_node.meta.items():
-                new_node.meta[k] = v
+    for node in gm.graph.nodes:
+        if node.op == "placeholder":
+            if node.target in new_graph_signature.inputs_to_parameters:
+                param_name = new_graph_signature.inputs_to_parameters[node.target]
+                if param_name in params_buffers_to_node_meta:
+                    for k, v in params_buffers_to_node_meta[param_name].items():
+                        node.meta[k] = v
+            if node.target in new_graph_signature.inputs_to_buffers:
+                buffer_name = new_graph_signature.inputs_to_buffers[node.target]
+                if buffer_name in params_buffers_to_node_meta:
+                    for k, v in params_buffers_to_node_meta[buffer_name].items():
+                        node.meta[k] = v
+
+    # overwrite signature for non-persistent buffers
+    for spec in new_graph_signature.input_specs:
+        if spec.kind == InputKind.BUFFER and spec.target in non_persistent_buffers:
+            spec.persistent = False
```

Comment:
Rewerite these for loops in to functions if possible? Ideally, the code could looks like the below _verify* calls.